src/pages/[platform]/build-a-backend/data/connect-to-existing-data-sources/single-table-design/index.mdx

```diff
@@ -0,0 +1,490 @@
+
+To interact with the external DynamoDB table, you can add custom types, mutations, and queries to the `amplify/data/resource.ts` file.
+
+Here's an example schema for the racing application:
+
+```ts
+import { type ClientSchema, a, defineData } from "@aws-amplify/backend";
+
+const schema = a.schema({
+  Race: a.customType({
+    raceId: a.string().required(),
+    classId: a.string(),
+    results: a.json(),
+  }),
+  RacingTable: a.customType({
+    pk: a.string().required(),
+    sk: a.string().required(),
+    numeric: a.float(),
+    results: a.json(),
+  }),
+  createRace: a
+    .mutation()
+    .arguments({
+      raceId: a.string().required(),
+      classId: a.string(),
+      results: a.json(),
+    })
+    .returns(a.ref("RacingTable"))
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./createRace.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+  getRaceResultsByRacerId: a
+    .query()
+    .arguments({
+      racerId: a.string().required(),
+    })
+    .returns(a.ref("RacingTable").array())
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./getRaceResultsByRacerId.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+  getRacesByClassId: a
+    .query()
+    .arguments({
+      classId: a.string().required(),
+    })
+    .returns(a.ref("RacingTable").array())
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./getRacesByClassId.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+  getBestPerformanceByClassId: a
+    .query()
+    .arguments({
+      classId: a.string().required(),
+    })
+    .returns(a.ref("RacingTable").array())
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./getBestPerformanceByClassId.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+});
+
+export type Schema = ClientSchema<typeof schema>;
+
+export const data = defineData({
+  schema,
+  name: "MyLibrary",
+  authorizationModes: {
+    defaultAuthorizationMode: "apiKey",
+    apiKeyAuthorizationMode: {
+      expiresInDays: 365,
+    },
+  },
+});
+```
+
+## Reviewing Data Access Patterns
+
+It's crucial to test the proposed schema against the list of data access patterns. Let's revisit the racing application's queries to ensure the table schema supports them:
+
+1. **Get results for each race by racer ID**: Use the table's partition key, searching for PK = racer ID. This returns a list of all races (PK) for a specific racer.
+
+<BlockSwitcher>
+<Block name="Resolver">
+```ts title="amplify/data/getRaceResultsByRacerId.js"
+import { util } from "@aws-appsync/utils";
+import * as ddb from "@aws-appsync/utils/dynamodb";
+
+export function request(ctx) {
+  const { racerId } = ctx.args;
+
+  return ddb.query({
```

Comment:
This might be useful from the amplify gen 2 docs, seems it's been done differently: https://docs.amplify.aws/react/build-a-backend/data/custom-business-logic/#step-2---configure-custom-business-logic-handler-code

Reply:
This deployed successfully for me and has the bonus of being typesafe. Under the hood I imagine @aws-appsync/utils/dynamodb is just using aws-sdk anyway:
```
// resourse.ts
...
const getMessageByConversationId = defineFunction({
  entry: './getMessagesByConversationId.ts'
})
...
  getMessagesByConversationId: a
    .query()
    .arguments({
      conversationId: a.string().required(),
      lastKey: a.string(),
    })
    .returns(a.ref("ChatTable").array())
    .handler(
      a.handler.function(getMessageByConversationId)
    )
    .authorization(allow => [allow.publicApiKey()]),
```

then my handler:
```
// getMessageByConversationId.ts

import { DynamoDB } from 'aws-sdk';
import type { Schema } from './resource';

const ddb = new DynamoDB.DocumentClient();

type GetMessagesByConversationIdResult = {
  pk: string;
  sk: string;
  type: string;
  message?: string | null;
  userId?: string | null;
  conversationId?: string | null;
  timestamp?: string | null;
}[];

export const handler: Schema["getMessagesByConversationId"]["functionHandler"] = async (
  event
): Promise<GetMessagesByConversationIdResult | null> => {
  const { conversationId, lastKey } = event.arguments;

  const queryInput: DynamoDB.DocumentClient.QueryInput = {
    TableName: "ChatTable",
    KeyConditionExpression: 'pk = :conversationId AND begins_with(sk, :prefix)',
    ExpressionAttributeValues: {
      ':conversationId': `CONVERSATION#${conversationId}`,
      ':prefix': 'MESSAGE#',
    },
    ScanIndexForward: true,
  };

  if (lastKey) {
    queryInput.ExclusiveStartKey = {
      pk: `CONVERSATION#${conversationId}`,
      sk: lastKey,
    };
  }

  try {
    const result = await ddb.query(queryInput).promise();

    // Transform the result.Items into the expected format
    const items: GetMessagesByConversationIdResult = (result.Items || []).map(item => ({
      pk: item.pk as string,
      sk: item.sk as string,
      type: item.type as string,
      message: item.message as string | null,
      userId: item.userId as string | null,
      conversationId: item.conversationId as string | null,
      timestamp: item.timestamp as string | null,
    }));

    return items;
  } catch (error) {
    console.error('Error processing request:', error);
    throw new Error('Failed to process request');
  }
};
```